File: sem iv/Theory of Computaion/lab/code/python/6_dfa_contains_101.py
# DFA accepting language over {0,1} such that string contains substring 101
# States: 0, 1, 2, 3 (accepting, trap)

def delta(current_state, symbol):
    if current_state == 0:
        return 0 if symbol == '0' else 1
    elif current_state == 1:
        return 2 if symbol == '0' else 1
    elif current_state == 2:
        return 0 if symbol == '0' else 3
    elif current_state == 3:
        return 3

def main():
    print("String has substring 101")
    w = input("Enter the input: ")
    current_state = 0
    for symbol in w:
        current_state = delta(current_state, symbol)
    if current_state == 3:
        print(f"{w} accepted")
    else:
        print(f"{w} rejected")

if __name__ == "__main__":
    main()

File: sem iv/Theory of Computaion/lab/code/c/6_dfa_contains_101.cpp
// DFA accepting language over {0,1} such that string contains substring 101
// States: q0, q1, q2, q3 (accepting, trap)
#include <iostream>
using namespace std;

int delta(int current_state, char symbol);

int main() {
    string w;
    cout << "String has substring 101" << endl;
    cout << "Enter the input: ";
    cin >> w;

    int current_state = 0;
    for (int i = 0; i < (int)w.length(); i++)
        current_state = delta(current_state, w[i]);

    if (current_state == 3)
        cout << w << " accepted" << endl;
    else
        cout << w << " rejected" << endl;
    return 0;
}

int delta(int current_state, char symbol) {
    switch (current_state) {
        case 0: current_state = (symbol == '0') ? 0 : 1; break;
        case 1: current_state = (symbol == '0') ? 2 : 1; break;
        case 2: current_state = (symbol == '0') ? 0 : 3; break;
        case 3: current_state = 3; break;
    }
    return current_state;
}
